bam_fastq.c

```diff
@@ -0,0 +1,1020 @@
+            case 'v': opts->def_qual = atoi(optarg); break;
+            case '?': bam2fq_usage(stderr, argv[0]); free_opts(opts); return false;
+            default:
+                if (parse_sam_global_opt(c, optarg, lopts, &opts->ga) != 0) {
+                    bam2fq_usage(stderr, argv[0]); free_opts(opts); return false;
+                }
+                break;
+        }
+    }
+
+    if (opts->fnr[1] || opts->fnr[2]) opts->has12 = false;
+    if (opts->has12always) opts->has12 = true;
+
+    if (!opts->barcode_tag) opts->barcode_tag = strdup(DEFAULT_BARCODE_TAG);
+    if (!opts->quality_tag) opts->quality_tag = strdup(DEFAULT_QUALITY_TAG);
+
+    int nIndex = 0;
+    if (opts->index_format) {
+        char *s;
+        for (s = opts->index_format; *s; s++) {
+            if (*s == 'i') nIndex++;
+        }
+    }
+    if (nIndex>2) {
+        fprintf(stderr,"Invalid index format: more than 2 indexes\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (opts->index_file[1] && !opts->index_file[0]) {
+        fprintf(stderr, "Index one specified, but index two not given\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (nIndex==2 && !opts->index_file[1]) {
+        fprintf(stderr, "index_format specifies two indexes, but only one index file given\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (nIndex==1 && !opts->index_file[0]) {
+        fprintf(stderr, "index_format specifies an index, but no index file given\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (nIndex==0 && opts->index_file[0]) {
+        fprintf(stderr, "index_format not specified, but index file given\n");
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    if (opts->def_qual < 0 || 93 < opts->def_qual) {
+        fprintf(stderr, "Invalid -v default quality %i, allowed range 0 to 93\n", opts->def_qual);
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    const char* type_str = argv[0];
+    if (strcasecmp("fastq", type_str) == 0 || strcasecmp("bam2fq", type_str) == 0) {
+        opts->filetype = FASTQ;
+    } else if (strcasecmp("fasta", type_str) == 0) {
+        opts->filetype = FASTA;
+    } else {
+        print_error("bam2fq", "Unrecognised type call \"%s\", this should be impossible... but you managed it!", type_str);
+        bam2fq_usage(stderr, argv[0]);
+        free_opts(opts);
+        return false;
+    }
+
+    int in_isnt_tty = !isatty(STDIN_FILENO);
+    if (in_isnt_tty && argc > optind && strcmp(argv[optind], "-") == 0)
+        in_isnt_tty = 0; // as we specified "-" explicitly anyway
+
+    if (argc - optind + in_isnt_tty == 0) {
+        fprintf(stderr, "No input file specified.\n");
```

Comment:
`samtools fastq` by itself is the canonical way to ask for the usage, so it shouldn't produce a _No input file specified_ error message (or return EXIT_FAILURE).

Reply:
The _No input file specified_ predated this PR, but it's now gone.  It should also now return EXIT_SUCCESS if just called as `samtools fastq` (but maybe not on Windows...)

I've also fixed a couple of dependencies in the Makefile that got lost when the fastq code was extracted from `sam_view.c`.

Reply:
Thanks × 2. Oops — it appeared on the GH web view to be new, but that's just because James moved the code to a new file, duh.